feat(folds): add 'foldoptions' option

andersevenrud opened the pull request:
Introduces a new option named 'foldoptions' for customizing the
fold column.

Comes with the word 'nodigits' that disables the digits shown to
indicate the nesting level. Instead 'foldsep' from 'fillchars' is
used.

---

I'm new to neovim hacking and very novice at C, so this is probably just one big mess... but I thought I would try to take a stab at implementing this option as I find the rendering of the nesting level as numbers quite disturbing over just showing the regular separator.

closes #14751

lewis6991 commented:
Hmm, `foldlevelmarker` sounds like it's related to `foldmarker`. Maybe `foldcolumnlevel` or `foldcolumndigits` would be beter?

lewis6991 commented on src/nvim/options.lua:
```diff
@@ -940,6 +940,14 @@ return {
+    {
+      full_name='foldlevelmarker', abbreviation='fdlm',
+      short_desc=N_("Enable digits shown to indicate the nesting level"),
+      type='bool', scope={'global'},
```
Shouldn't this be `window` scoped like other fold options?

andersevenrud replied:
That's how I implemented this at first (and which I can revert to), after some testing and contemplation I decided to go this way because because it feels like it should be a global setting since it's related to the gutter visuals.

andersevenrud replied:
You're probably right here. So I reverted back to window scoped.

lewis6991 replied:
Yeah was gonna say `'signcolumn'` and `'number'` are window scoped too.

andersevenrud commented:
> Maybe foldcolumnlevel or foldcolumndigits would be beter?

Ah, yes. I was hoping someone suggested something else. I felt really unsure about the naming.

lewis6991 commented:
I'm not an expert in this area of the codebase but LGTM :+1:

zeertzjq commented:
I think this should now be put after `'foldcolumn'` instead of after `'foldlevel'`.

andersevenrud commented:
@zeertzjq Thanks for the suggestion. I've moved it.

andersevenrud commented on src/nvim/options.lua:
```diff
@@ -940,6 +940,13 @@ return {
       defaults={if_true="foldtext()"}
     },
+    {
+      full_name='foldcolumndigits', abbreviation='fdcd',
```
Fixed.

andersevenrud commented:
Not really sure why the tests failed (to start?).

They run successfully locally at least.

clason commented:
At first glance, this looks like the usual GHA flake. I restarted the tests.

clason commented:
And now it passed :)

Is the plan to submit this to Vim as well? If not, this should probably be added to `vim-diff.txt`:https://github.com/neovim/neovim/blob/36362ef0aed92e726d967d30e3c6cd87c65642b3/runtime/doc/vim_diff.txt#L226-L240

andersevenrud commented:
> Is the plan to submit this to Vim as well?

I will definitively give it a go when I have some time.

In the meantime, I'll update the `vim-diff.txt` documentation as suggested.

justinmk reviewed (changes requested):
a new single-purpose option is not the right approach. either consider a sub-flag in the existing `foldcolumn` option, or maybe introduce some other more generic mechanism, perhaps a callback that controls rendering. Probably a new `foldoptions` option is the right pattern here, similar to 'wildoptions', 'jumpoptions', etc.

lewis6991 commented:
~@justinmk: I see your point, but I half-disagree. `foldcolumn` really just defines how much space the fold column occupies ...~

Just saw your edit whilst composing a response. 👍 to adding `foldoptions`. I don't think overloading `foldcolumn` is the right approach considering that it mostly mirrors `signcolumn`.

xiyaowong commented:
Is there any further work on this pr, it's a simple but very useful feature :lipstick:

andersevenrud commented:
I've been very busy since I initially created this PR, so haven't had a chance to look into the feedback / requests.

But I now finally have some time to spare, so I'll try to implement @justinmk's suggestion here ASAP.

justinmk commented:
For reference, `jumpoptions` was a similar case where we anticipated future options, and we now have 2 flags there. So introducing `foldoptions` makes sense for these kinds of on/off flags.

andersevenrud commented:
> Should be an easy reference to add 'foldoptions'.

Thanks a lot for the pointers @justinmk . I gave this a go and it seems like things are working as it should.

sunzoje commented on runtime/doc/vim_diff.txt:
```diff
@@ -235,6 +235,7 @@ Options:
   'fillchars'   flags: "msgsep" (see 'display'), "horiz", "horizup",
                 "horizdown", "vertleft", "vertright", "verthoriz"
   'foldcolumn'  supports up to 9 dynamic/fixed columns
+  'foldoptions' fold colum customization options
```
```suggestion
  'foldoptions' fold column customization options
```

andersevenrud replied:
Thanks! Fixed :)

lewis6991 commented on runtime/doc/options.txt:
```diff
@@ -2650,6 +2650,12 @@ A jump table for the options with a short description can be found at |Q_op|.
+'foldoptions' 'fop'	string (default "")
+			local to window
+	List of words that change the behavor of the |foldcolumn|.
+          nodigits      Disable digits shown to indicate the nesting level.
```
I would change this to `digits` and add it to the default.

andersevenrud replied:
I've tried doing this but can't seem a find a way to properly set the default value. Doing it in the lua options definitions has no effect.

Maybe you have some pointers on this?

justinmk replied:
> I've tried doing this but can't seem a find a way to properly set the default value. Doing it in the lua options definitions has no effect.

did you look at existing examples, e.g. https://github.com/neovim/neovim/blob/9370e1c5111ee90f64260398b0623da4759f8f16/src/nvim/options.lua#L633-L640

and what exactly did you try

andersevenrud replied:
@justinmk 

Yeah, I've looked hard at finding a solution to my issue. But totally hit a wall.

Here's my uncomitted work based on this branch:

```patch
diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index 536cce727..6a81e5dc2 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -2651,10 +2651,10 @@ A jump table for the options with a short description can be found at |Q_op|.
        set the 'foldclose' option to "all".

                                                      *'foldoptions'* *'fop'*
-'foldoptions' 'fop'    string (default "")
+'foldoptions' 'fop'    string (default "digits")
                        local to window
-       List of words that change the behavor of the |foldcolumn|.
-          nodigits      Disable digits shown to indicate the nesting level.
+       List of words that change the behavor of the |folding|.
+          digits        Enable digits shown to indicate the nesting level.

                                                *'foldtext'* *'fdt'*
 'foldtext' 'fdt'       string (default: "foldtext()")
diff --git a/src/nvim/option_defs.h b/src/nvim/option_defs.h
index a4de28f2c..36d1e9c0f 100644
--- a/src/nvim/option_defs.h
+++ b/src/nvim/option_defs.h
@@ -461,9 +461,9 @@ static char *(p_fdo_values[]) = { "all", "block", "hor", "mark", "percent",
 #define FDO_UNDO               0x200
 #define FDO_JUMP               0x400
 #ifdef IN_OPTION_C
-static char *(p_fop_values[]) = { "nodigits", NULL };
+static char *(p_fop_values[]) = { "digits", NULL };
 #endif
-#define FOP_NODIGITS           0x01
+#define FOP_DIGITS             0x01
 EXTERN char_u *p_fp;          // 'formatprg'
 EXTERN int p_fs;                // 'fsync'
 EXTERN int p_gd;                // 'gdefault'
diff --git a/src/nvim/options.lua b/src/nvim/options.lua
index 8bb4ad29d..afdb127bb 100644
--- a/src/nvim/options.lua
+++ b/src/nvim/options.lua
@@ -867,7 +867,7 @@ return {
       type='string', list='onecomma', scope={'window'},
       deny_duplicates=true,
       redraw={'current_window'},
-      defaults={if_true=''}
+      defaults={if_true='digits'}
     },
     {
       full_name='foldenable', abbreviation='fen',
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index 7562326f5..c73b195c6 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -1922,7 +1922,7 @@ static size_t fill_foldcolumn(char_u *p, win_T *wp, foldinfo_T foldinfo, linenr_
   int symbol = 0;
   int len = 0;
   bool closed = foldinfo.fi_lines > 0;
-  bool no_digits = wp->w_p_fop_flags & FOP_NODIGITS;
+  bool no_digits = !(wp->w_p_fop_flags & FOP_DIGITS);
   // Init to all spaces.
   memset(p, ' ', MAX_MCO * fdc + 1);

diff --git a/test/functional/ui/fold_spec.lua b/test/functional/ui/fold_spec.lua
index f73459038..f40999cfe 100644
--- a/test/functional/ui/fold_spec.lua
+++ b/test/functional/ui/fold_spec.lua
@@ -1866,7 +1866,7 @@ describe("folded lines", function()
         ]])
       end

-      command('set foldoptions=nodigits')
+      command('set foldoptions=')
       if multigrid then
         screen:expect([[
           ## grid 1
```

It doesn't seem like `defaults` does anything when the scope is set to `window`, unless I'm totally missing something obvious here :thinking:

andersevenrud replied:
I studied the `cursorlineopt` option (and quite a few others), but it didn't really get me anywhere :(

> Maybe try redraw={'current_window_only'} (just a guess) ?

This has no effect.

lewis6991 commented on runtime/doc/options.txt:
```diff
@@ -2650,6 +2650,12 @@ A jump table for the options with a short description can be found at |Q_op|.
+    						      *'foldoptions'* *'fop'*
+'foldoptions' 'fop'	string (default "")
+			local to window
+	List of words that change the behavor of the |foldcolumn|.
```
I think this option can be made general to folding, not just the fold column.

teto commented:
In favor of a fold-related option too, I wanted to introduce one https://github.com/neovim/neovim/pull/13617 to toggle column-based folds. I am sure it will be useful in the future

justinmk commented:
# Status

- https://github.com/neovim/neovim/pull/17446#discussion_r912447977 is not a blocker (though it does smell like tech debt and would be great if someone can look into it)
- just need a rebase to fix conflicts then we can merge this

p00f commented:
@andersevenrud I rebased it for you (I'm eager to get rid of the digits), here's a patch on top of current master `d5cc161deac2371a39821c94fd2de4f1c41b6eb9` - https://paste.sr.ht/~p00f/d575eeda2986cc0234a6e3dd718093992f314662

p00f commented:
@andersevenrud I rebased it and made a patch for you here

https://paste.sr.ht/~p00f/c534a0b3ae75ed4afdc87f8f1870d971acf3f9b4


String options were moved to optionstr.c so I'm not sure if adding a guard like this was right
https://paste.sr.ht/~p00f/c534a0b3ae75ed4afdc87f8f1870d971acf3f9b4#0001-feat-folds-add-foldoptions-option.patch-L170
https://paste.sr.ht/~p00f/c534a0b3ae75ed4afdc87f8f1870d971acf3f9b4#0001-feat-folds-add-foldoptions-option.patch-L128

zeertzjq commented:
Put `p_fop_values` in `optionstr.c` instead.

p00f commented:
Like this? https://paste.sr.ht/~p00f/380788a5051de7ee94e315feb98e1f1016a9d3fa

zeertzjq commented:
Yes.

andersevenrud committed:
feat(folds): add 'foldoptions' option

Introduces a new option named 'foldoptions' for customizing the
fold column.

Comes with the word 'nodigits' that disables the digits shown to
indicate the nesting level. Instead 'foldsep' from 'fillchars' is
used.

andersevenrud commented:
@p00f Really appreciate it :1st_place_medal:  So much stuff going on in life right now that I would not have found time for this in quite a while :sweat: 

This PR has had your patch applied.

p00f commented:
There is one issue though - when I open a new window the foldcolumn shows the numbers. I need to set the option again, even though doing `:set fop?` shows `nodigits` in that window. For the time being I got around it using a `WinNew` autocommand

andersevenrud commented:
> There is one issue though - when I open a new window the foldcolumn shows the numbers

This is intended. It's per-window setting.

> I need to set the option again, even though doing :set fop? shows nodigits in that window

But this ain't. I can confirm that you'll see nodigits set even if not active.

zeertzjq commented:
> This is intended. It's per-window setting.

No this is not. Window-local settings should be properly copied when splitting the window.

zeertzjq commented on src/nvim/buffer_defs.h:
```diff
@@ -1327,6 +1329,7 @@ struct window_S {
   uint32_t w_p_wbr_flags;           // flags for 'winbar'
   uint32_t w_p_fde_flags;           // flags for 'foldexpr'
   uint32_t w_p_fdt_flags;           // flags for 'foldtext'
+  uint32_t w_p_fop_flags;           // flags for 'foldoptions'
```
Move to `winopt_T` as `wo_fop_flags`. And `uint8_t` is enough for these flags.

justinmk commented:
> new option named 'foldoptions' for customizing the fold column.

Perhaps this isn't needed now that we have 'statuscolumn'. https://github.com/neovim/neovim/pull/20621

towry commented:
How to hide the fold level number with `statuscolumn`? As I know, many people were waiting for this PR to be merged.

https://github.com/kevinhwang91/nvim-ufo/issues/4#issuecomment-1157716294

luukvbaal commented:
Nice! I will look into adding support for things like this to [statuscol.nvim](https://github.com/luukvbaal/statuscol.nvim) as well.

p00f commented:
> Not a vimscript expert but a quick attempt. Does not use existing fillchars option. Just inserts a chevron for open and closed folds.

> `vim.o.statuscolumn = '%=%l%s%{foldlevel(v:lnum) > foldlevel(v:lnum - 1) ? (foldclosed(v:lnum) == -1 ? "▼" : "⏵") : " " }'`

To show the vertical bar thing (ie same as default vim except the numbers)
`vim.o.statuscolumn='%=%l%s%{foldlevel(v:lnum) > 0 ? (foldlevel(v:lnum) > foldlevel(v:lnum - 1) ? (foldclosed(v:lnum) == -1 ? "-" : "+") : "|") : " " }'`

rockyzhang24 commented:
> vim.o.statuscolumn = '%=%l%s%{foldlevel(v:lnum) > foldlevel(v:lnum - 1) ? (foldclosed(v:lnum) == -1 ? "▼" : "⏵") : " " }'

This is just a workaround to simulate the fold signs by an expression, so actually it is not a real foldcolumn. This results in the failure of the `FoldColumn` highlight group. We have to specify it explicitly via `%#FoldColumn#%{ ... }%*` as `statusline` does.

So I think this PR is still necessary to tweak the foldcolumn directly. If so, we could just use `statuscolumn` to arrange items instead of introducing expressions to tweak items, which will be more efficient and reduct the potential performance overhead. 

What do you think @justinmk ? Thank you so much.

kevinhwang91 commented:
> vim.o.statuscolumn = '%=%l%s%{foldlevel(v:lnum) > foldlevel(v:lnum - 1) ? (foldclosed(v:lnum) == -1 ? "▼" : "⏵") : " " }'

buggy.

```lua
if true then
    print('true')
else
    print('false')
end
````
Consider `1,2fold | 3,4fold | norm! zR`. May use FFI to invoke `foldinfo_T fold_info(win_T *win, linenr_T lnum);` to get `fi_low_level`

Should reopen this PR.

justinmk commented:
This PR in its current form is too special-case. Need to see a proposal showing a more complete picture of what parts of fold UI need configuration. Adding booleans here and there like "nodigits" is very narrowly focused. 

> So I think this PR is still necessary to tweak the foldcolumn directly.

This PR only adds "nodigits". How does that allow you to tweak the foldcolumn directly?

rockyzhang24 commented:
> This PR only adds "nodigits". How does that allow you to tweak the foldcolumn directly?

"tweak" is no matter how big or small 😁. At least we could toggle showing the digits or not in foldcolumn. Yes, I totally agree that creating a new option for this kinda special-case is not good choice. Probably we could integrate `nodigits` as a new value with an existing option like `foldcolumn`, e.g., `vim.o.foldcolumn = 'nodigits'`.

luukvbaal commented:
> I will look into adding support for things like this to [statuscol.nvim](https://github.com/luukvbaal/statuscol.nvim) as well.

So it took some time to get around to this(I don't really use folds myself). I do think there are gaps in the current api to properly emulate the fold column through `'statuscolum'`. As @kevinhwang91 mentioned, we need the information in `foldinfo_T`. Would a builtin `foldinfo({winid}, {lnum})` function be welcome @justinmk to achieve this? Using that, I was able to emulate a "nodigits" fold column like so:
```lua
function M.foldfunc()
	local fi = vim.fn.foldinfo(v.windowid, v.lnum)
	local width = fi.maxdepth
	if width == 0 then return "" end

	local string = "%#FoldColumn#"
	local level = fi.level

	if level == 0 then
		return string..(" "):rep(width).."%#LineNr#"
	end

	-- For each column, add a foldopen, foldsep or whitespace char
	for col = 1, width do
		if fi.start == v.lnum and col >= fi.low_level then
			string = string..foldopen
		else
			string = string..foldsep
		end

		if col == level then
			string = string..(" "):rep(width - col)
			break
		end
	end

	-- If line contains a closed fold, substitute foldclosed char
	if fi.lines > 0 then
		local pos = string:len()
		if width >= level then pos = pos - (width - level) end
		string = string:sub(1, pos - 1)..foldclosed..string:sub(pos + 1)
	end

	return string.."%#LineNr#"
end
```

> So I think this PR is still necessary to tweak the foldcolumn directly. If so, we could just use statuscolumn to arrange items instead of introducing expressions to tweak items, which will be more efficient and reduct the potential performance overhead.

Although I have no objections to re-opening this PR to tweak the default column directly, performance concerns for the above approach seem to be a non-issue. At least I notice no discernible performance hit with the above approach and the following fold column(default fold column on the right split):
![image](https://user-images.githubusercontent.com/31730729/214998589-4bef7e5a-adad-4fd5-8706-4c3d6d2df78b.png)

rockyzhang24 commented:
@luukvbaal Good work. Thank you for digging into the problem and figuring out the workaround.

>  At least I notice no discernible performance hit

I am not sure but the performance overhead may be noticeable if we have hundreds of code lines in one screen (e.g., using a monitor in portrait orientation), especially on some old low-performance machines.

luukvbaal commented:
[statuscol.nvim](https://github.com/luukvbaal/statuscol.nvim) now supports a "nodigits" foldcolumn through FFI:
```lua
require("statuscol").setup({ foldfunc = "builtin", setopt = true })
```

ReStranger commented:
> [statuscol.nvim ](https://github.com/luukvbaal/statuscol.nvim) теперь поддерживает столбец сгиба «без цифр» через FFI:
> 
> ```lua
> local builtin = require("statuscol.builtin")
> require("statuscol").setup({
>   segments = {
>     { text = { "%s" }, click = "v:lua.ScSa" },
>     { text = { builtin.lnumfunc }, click = "v:lua.ScLa", },
>     {
>       text = { " ", builtin.foldfunc, " " },
>       condition = { builtin.not_empty, true, builtin.not_empty },
>       click = "v:lua.ScFa"
>     },
> }
> ```

For some reason it still displays the numbers.  Can anyone help me?
